### PR TITLE
reorg of brts to separate part-tests

### PR DESCRIPTION
**Pull request by jerinphilip**
There are two kind of BRT tests intended:

1. Exact: Where exact outputs are used for regression-tests. Probably for use in var, where all things are controllable and consistent across runs. 
2. Approximate: Where we require that outputs only approximately match expected outputs (BLEU, numerics with some tolerance for precision etc). For use in GitHub CI and across developer machines.

As a switch, a shell variable is used where someone specifies "approx" or "exact". If approx, we evaluate against with avx512vnni (the approx directory is symlinked to this) that the outputs "approximately" match to the expected output. 

```
MARIAN=../build BRT_EVAL_MODE=approx ./run-brt.sh
MARIAN=../build BRT_EVAL_MODE=exact ./run-brt.sh

```

There are cases two implementations of approx in mind:
1. Approximate text-match: Using BLEU here, --sentence-only option, also --greater-than `<value in [0, 100]>`. 
2. Approximate number-match: Using `diff-nums.py`, from marian regression-tests here allowing a few faults.

If I retain individual arch outputs things seem mostly okay (Only precision issues, save very few faults). This is if we retain `gemm-precision x instruction` matrix as it stands now. 

Across archs, there seem to be difference in outputs created by the difference in floating point operations and a sampling something in between which creates the difference in elements being compared which makes approximate tests go haywire as of now.

### CODE REVIEW

File: tools/resolve-instruction.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    
+    commands = [
+        'export {}={}'.format(key, value) for key, value in envvars.items()
+    ]
+    return ('\n'.join(commands))
+
+def main(args):
+    # Run `list_cpu_features` to obtain JSON output.
+    proc = sp.run(args=[args.path, '--json'], stdout=sp.PIPE, stderr=sp.PIPE)
```
Comment (kpu) [inline tools/resolve-instruction.py:67]: So far as I can tell, this is a Rube Goldberg device to get the running CPU model by way of:
- A third party library https://github.com/google/cpu_features#running-sample-code
- Some JSON file
- And some python program that parses the JSON

We already have CPUID, just print out `intgemm::kCPU`  And use a bash case statement to map to environment variables.  

```
// If you want to detect the CPU and dispatch yourself, here's what to use:
enum class CPUType {
  UNSUPPORTED = 0,
  SSE2 = 1,
  SSSE3 = 2,
  AVX2 = 3,
  AVX512BW = 4,
  AVX512VNNI = 5
};
```

File: tools/approx-diff.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            nonEmpty = lambda x: len(x) > 0
+            content = list(filter(nonEmpty, content))
+        return content
+
+    system = load_non_empty_lines(args.file1)
+    refs = [ load_non_empty_lines(args.file2) ]
+
+    faults = 0
+    max_allowed_faults = 0
+    if args.sentence_level:
```
Comment (kpu) [inline tools/approx-diff.py:62]: Why do we want sentence bleu?

**Review by kpu — CHANGES_REQUESTED** [review]
I'll have more comments.

File: .gitignore [hunk 3]
```diff
@@ -16,6 +16,7 @@ tools/moses-scripts
 tools/subword-nmt
 tools/nematus
 tools/sacrebleu
+tools/cpu-features
```
Comment (kpu) [inline .gitignore:19]: Remove this.

**Review by kpu — CHANGES_REQUESTED** [review]
Why do we have fixed outputs for each arch when a more sane design was just a test set (of say 3000 sentences) and bleu on it, without regard to arch?

**Review by kpu — CHANGES_REQUESTED** [review]
BLEU doesn't make sense for evaluating alignments.  Anyway the alignments are not useful in their present form (they should be a dense matrix that will then go to tag handling), so it's not particularly useful to be making regression tests for unused output anyway.

File: tools/resolve-instruction.py (hunk 1 above)
Comment (jerinphilip) [reply tools/resolve-instruction.py]: This change is in effect, except I do not know what to set as MKL vars for SSSE3 and SSE2, which are supported by intgemm, but unable to find an equivalent for MKL. 

There is no information regarding SSSE3 or SSE2 here: https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/support-functions/miscellaneous/mkl-enable-instructions.html.

Docs elsewhere say just change ISA to MKL_ENABLE_INSTRUCTIONS=ISA , but ISA=SSSE3,SSE2 is not available.

File: tools/approx-diff.py (hunk 2 above)
Comment (jerinphilip) [reply tools/approx-diff.py:62]: I simply tried sentence-bleu to see if that works. Given how the scripts are (over)-engineered, if someone supplies a `check`, where

```
check OUTPUT_TEXT_FILE EXPECTED_TEXT_FILE
```

evaluates if an output approximately matches expected output, it should be super simple to replace.

**Comment by jerinphilip** [comment]
> Why do we have fixed outputs for each arch when a more sane design was just a test set (of say 3000 sentences) and bleu on it, without regard to arch?

The objective is to ensure that the constructed `Response` out of the translation pipeline is intact and not affected by changes by an incoming PR. This means we're at least checking for breakages in:

1. translated text, where faithful translation feature **requires exact match**, not approximate like BLEU.
2. Sentences API, where we can possibly downgrade to BLEU based evaluations.
3. Token level annotations, where we **require exact matches** to know if something broke/changed between development. BLEU will let minor changes slide here. 

What remains is alignments and quality scores.

> BLEU doesn't make sense for evaluating alignments. Anyway the alignments are not useful in their present form (they should be a dense matrix that will then go to tag handling), so it's not particularly useful to be making regression tests for unused output anyway.

Given all the headhurt, I'm down for removing quality-scores and alignments than put any extra effort.  However, at the time I was put on doing this, I was led to believe that we were supposed to supply the sparse matrix and extension the tag handling.

**Comment by kpu** [comment]
Let's move on to more exciting features

**Pull request merged**